internal/normalizer/normalizer.go

```diff
@@ -0,0 +1,28 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package normalizer provides tag normalization
+package normalizer
+
+import (
+	"regexp"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+)
+
+// NormalizeHeaderTag accepts a string that contains a header and an optional mapped tag key,
+// e.g, "header" or "header:tag" where `tag` will be the name of the header tag.
+func NormalizeHeaderTag(headerAsTag string) (header string, tag string) {
+	lastIdx := strings.LastIndex(headerAsTag, ":")
+
+	// If no colon or colon sits at the very beginning or very end of the string
+	if lastIdx == -1 || lastIdx == 0 || lastIdx == len(headerAsTag)-1 {
+		headerAsTag = strings.ToLower(strings.TrimSpace(headerAsTag))
+		regex := regexp.MustCompile(`[^a-zA-Z0-9 -]+`)
+		return headerAsTag, ext.HTTPRequestHeaders + "." + regex.ReplaceAllString(headerAsTag, "_")
+	}
+	return strings.ToLower(strings.TrimSpace(headerAsTag[:lastIdx])), strings.ToLower(strings.TrimSpace(headerAsTag[lastIdx+1:]))
```

Comment:
Suggested refactor for readability: https://go.dev/play/p/MczD7Lshfz9
  - If you know it's going to have to be lowercase, then just make the entire string lowercase at the beginning of the function. i.e. first line is `headerAsTag = strings.ToLower(headerAsTag)`. Then you don't need to call it several times in this fn, and you can drop the capital letters from the regexp
  - I found the if statement on the index a bit difficult to parse, and some of the lines a bit too long.

```
func normalize(headerAsTag string) (header string, tag string) {
	header = strings.ToLower(strings.TrimSpace(headerAsTag))
	if last := strings.LastIndex(header, ":"); last >= 0 {
		// Split the tag from the colon if one was found.
		header, tag = header[:last], header[last+1:]
		header, tag = strings.TrimSpace(header), strings.TrimSpace(tag)
		tag = headerTagRegexp.ReplaceAllString(tag, "_") // replace invalid characters
	} else {
		tag = fmt.Sprintf("http.%s", headerTagRegexp.ReplaceAllString(header, "_"))
	}
	return header, tag
}
```

Reply:
@katiehockman , I specifically did not perform the lowercasing before the parsing because of this statement in the spec: 
> NB: if the user specifies a mapping part cf [above configuration 2)](https://datadoghq.atlassian.net/wiki/spaces/APMINT/pages/2302444638/DD+TRACE+HEADER+TAGS#mapping-the-header), no normalization should happen. The user is responsible for the mapping part, if some nesting is created unintentionally then configuration shoudl be fixed.

I took this to mean that if a user includes a mapping part, that this should remain untouched (except for removal of leading/trailing whitespaces after the colon & before the end of the string). LMK if you still think that lowercasing should join whitespacing though (Or alternatively that "0 normalization" = truly nothing, i.e, _no_ trimming of whitespaces?)

Reply:
TODO: Run parametric tests to check how other libraries handle this.

Reply:
@katiehockman , in my original logic, I was handling the following cases as if they include no tag mapping / as if the colon was merely a mistake:
`:header`
`header:`
That is, I was treating both cases as if the input was simply "header", no tag mapping specified.

But the logic you provided in this thread treats the above as input with valid tag mapping.
- For input ":header", header "" gets mapped to tag "header"
- For input "header:", header "header" gets mapped to tag ""

I can also try to use the parametric tests to check how the other libraries handle this input, because it is up for debate at this point.